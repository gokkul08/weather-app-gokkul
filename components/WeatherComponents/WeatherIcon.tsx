import React from 'react';
import Image from 'next/image';


interface WeatherIconProps {
    iconId: string;
};

const WeatherIcon = (props: WeatherIconProps) => {

    const { iconId } = props;

    return <Image src={`https://openweathermap.org/img/w/${iconId}.png`} width={50} height={50}/>
};

export default WeatherIcon;