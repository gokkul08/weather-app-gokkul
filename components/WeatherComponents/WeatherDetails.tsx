import React, { useState } from 'react';
import styled from 'styled-components';
import Container from '@material-ui/core/Container';
import Grid from '@material-ui/core/Grid';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import Paper from '@material-ui/core/Paper';
import WeatherIcon from './WeatherIcon';
import Switch from '@material-ui/core/Switch';
import Typography from '@material-ui/core/Typography';
import { timeFormatter, temperatureInCelcius, temperatureInFaranheit } from '../../utilities/weatherUtilities';

const StyledContainer = styled(Container)`
        padding-top: 20px;
    `;

interface WeatherDetailsProps {
    data: {
        daily: [
            {
                dt: number, 
                temp: {day: number, night: number}, 
                sunrise: number, 
                sunset: number, 
                moonrise: number, 
                moonset: number,
                weather: [{icon: string}],
                humidity: number,
            }
        ],
        hourly: [
            {
                dt: number, 
                temp: number, 
                feels_like: number, 
                weather: [{icon: string}],
            }
        ],
        timezone_offset: number,
    };
};

const WeatherDetails = (props: WeatherDetailsProps) => {
    const { data } = props;

    const [state, setState] = useState(false);

    const handleTemperatureChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        setState(event.target.checked);
    };

    const createWeeklyData = (week: string, day: string, temperatureday: number, temperaturenight: number, sunrise: string, sunset: string, moonrise: string, moonset: string, humidity: number, iconId: string) => {
        return { week, day, temperatureday, temperaturenight, sunrise, sunset, moonrise, moonset, humidity, iconId };
    };

    const createDailyData = (time: string, temperature: number, feels_like: number, iconId: string) => {
        return { time, temperature, feels_like, iconId };
    };

      
    const weeklyRows = data.daily.map((item) => {
        const day = timeFormatter(item.dt, data.timezone_offset);
        const temperatureday = state ? temperatureInCelcius(item.temp.day) : temperatureInFaranheit(item.temp.day);
        const temperaturenight = state ? temperatureInCelcius(item.temp.night) : temperatureInFaranheit(item.temp.night);
        const sunrise = timeFormatter(item.sunrise, data.timezone_offset);
        const sunset = timeFormatter(item.sunset, data.timezone_offset);
        const moonrise = timeFormatter(item.moonrise, data.timezone_offset);
        const moonset = timeFormatter(item.moonset, data.timezone_offset);
        const humidity = item.humidity;
        const iconId = item.weather[0].icon;
        return createWeeklyData(`${day.day}`, `${day.month},${day.date}`, temperatureday, temperaturenight, `${sunrise.hours}:${sunrise.minutes}`, `${sunset.hours}:${sunset.minutes}`, `${moonrise.hours}:${moonrise.minutes}`, `${moonset.hours}:${moonset.minutes}`, humidity, iconId);
    });

    const dailyRows = data.hourly.map((item) => {
        const time = timeFormatter(item.dt, data.timezone_offset);
        const temperature = state ? temperatureInCelcius(item.temp) : temperatureInFaranheit(item.temp);
        const feels_like = state ? temperatureInCelcius(item.feels_like) : temperatureInFaranheit(item.feels_like);
        const iconId = item.weather[0].icon;
        return createDailyData(`${time.hours}:${time.minutes}`, temperature, feels_like, iconId);
    });

    return (
        <StyledContainer maxWidth="md">
            <Grid container spacing={2}>
                    <Grid item xs={12}>
                        <b>&nbsp;°F</b>
                        <Switch 
                            onChange={handleTemperatureChange}
                            color="default"
                        />
                        <b>°C</b>
                        <Typography>
                            Weather Forcast for the week
                        </Typography>
                    </Grid>
                    <Grid item xs>
                        <TableContainer component={Paper}>
                            <Table>
                                <TableHead>
                                <TableRow>
                                    <TableCell align="right"><b>Week</b></TableCell>
                                    <TableCell align="right"><b>Date</b></TableCell>
                                    <TableCell align="right"><b>Day</b></TableCell>
                                    <TableCell align="right"><b>Night</b></TableCell>
                                    <TableCell align="right"><b>Sunrise</b></TableCell>
                                    <TableCell align="right"><b>Sunset</b></TableCell>
                                    <TableCell align="right"><b>Moonrise</b></TableCell>
                                    <TableCell align="right"><b>Moonset</b></TableCell>
                                    <TableCell align="right"><b>Humidity</b></TableCell>
                                    <TableCell align="right"><b>Forecast</b></TableCell>
                                </TableRow>
                                </TableHead>
                                <TableBody>
                                {weeklyRows.map((row, index) => (
                                    <TableRow key={index}>
                                        <TableCell align="right">{row.day}</TableCell>
                                        <TableCell align="right">{row.week}</TableCell>
                                        <TableCell align="right">{row.temperatureday}</TableCell>
                                        <TableCell align="right">{row.temperaturenight}</TableCell>
                                        <TableCell align="right">{row.sunrise}</TableCell>
                                        <TableCell align="right">{row.sunset}</TableCell>
                                        <TableCell align="right">{row.moonrise}</TableCell>
                                        <TableCell align="right">{row.moonset}</TableCell>
                                        <TableCell align="right">{row.humidity}</TableCell>
                                        <TableCell align="right"><WeatherIcon iconId={row.iconId} /></TableCell>
                                    </TableRow>
                                ))}
                                </TableBody>
                            </Table>
                        </TableContainer>
                    </Grid>
                    <Grid item xs>
                    <Typography>
                            Weather Forcast for the week
                        </Typography>
                        <TableContainer component={Paper}>
                                <Table>
                                    <TableHead>
                                    <TableRow>
                                        <TableCell align="right"><b>Time</b></TableCell>
                                        <TableCell align="right"><b>Temperature</b></TableCell>
                                        <TableCell align="right"><b>Feels Like</b></TableCell>
                                        <TableCell align="right"><b>Forecast</b></TableCell>
                                    </TableRow>
                                    </TableHead>
                                    <TableBody>
                                    {dailyRows.map((row, index) => (
                                        <TableRow key={index}>
                                            <TableCell align="right">{row.time}</TableCell>
                                            <TableCell align="right">{row.temperature}</TableCell>
                                            <TableCell align="right">{row.feels_like}</TableCell>
                                            <TableCell align="right"><WeatherIcon iconId={row.iconId} /></TableCell>
                                        </TableRow>
                                    ))}
                                    </TableBody>
                                </Table>
                            </TableContainer>
                    </Grid>
            </Grid>
        </StyledContainer>
    );
};

export default WeatherDetails;