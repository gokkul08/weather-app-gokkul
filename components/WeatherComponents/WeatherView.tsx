/**
 * This component provides the weather view after we complete the search for a location
 * It accepts the weather data as props, manipulates the data to make it available for viewing on the screen
 */

import React, { useState } from 'react';
import Grid from '@material-ui/core/Grid';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import CardActions from '@material-ui/core/CardActions';
import Typography from '@material-ui/core/Typography';
import WeatherIcon from './WeatherIcon';
import Switch from '@material-ui/core/Switch';
import Button from '@material-ui/core/Button';
import Link from 'next/link';
import { timeFormatter, temperatureInCelcius, temperatureInFaranheit } from '../../utilities/weatherUtilities';

interface WeatherViewProps {
    data: {
        coord: {
            lat: number,
            lon: number,
        }
        dt: number,
        main: {
            temp: number, 
            temp_max: number,
            temp_min: number,
            feels_like: number,
            pressure: number,
            humidity: number,
        },
        name: string,
        weather: [
            {
                main: string,
                icon: string,
            }
        ],
        sys: {
            sunrise: number,
            sunset: number,
            country: string,
        }
        timezone: number,
        wind: {
            deg: number,
            speed: number,
        }
    };
};

export const WeatherView = (props: WeatherViewProps) => {
    const { data } = props;

    const [state, setState] = useState(false);

    const latitute = data.coord && data.coord.lat;
    const longitude = data.coord && data.coord.lon;
    const currentTime = timeFormatter(data.dt, data.timezone);
    const sunRiseTime = timeFormatter(data.sys.sunrise, data.timezone);
    const sunSetTime = timeFormatter(data.sys.sunset, data.timezone);
    const currentTemperature = state ? temperatureInCelcius(data.main.temp) : temperatureInFaranheit(data.main.temp);
    const maxTemperature = state ? temperatureInCelcius(data.main.temp_max) : temperatureInFaranheit(data.main.temp_max);
    const minTemperature = state ? temperatureInCelcius(data.main.temp_min) : temperatureInFaranheit(data.main.temp_min);
    const feelsLikeTemperature = state ? temperatureInCelcius(data.main.feels_like) : temperatureInFaranheit(data.main.feels_like);
    const locationName = `${data.name}, ${data.sys.country}`;
    const weatherDescription = data.weather[0].main;
    const windDegree = data.wind.deg;
    const windSpeed = data.wind.speed;
    const humidity = data.main.humidity;
    const pressure = data.main.pressure;
    const iconId = data.weather[0].icon;

    const handleTemperatureChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        setState(event.target.checked);
    };

    return (
        <>
            <Grid container spacing={2}>
                <Grid item xs={8}>
                    <Card>
                        <Grid container spacing={2}>
                            <Grid item xs={12}>
                                    <b>&nbsp;°F</b>
                                <Switch 
                                    onChange={handleTemperatureChange}
                                    color="default"
                                />
                                    <b>°C</b>
                            </Grid>
                            <Grid item xs={4}>
                                <CardContent>
                                    <Typography variant="h3">
                                        {currentTemperature}°
                                    </Typography>
                                    <Typography>
                                        Feels like <b>{feelsLikeTemperature}°</b>
                                    </Typography>
                                    <WeatherIcon iconId={iconId} />
                                </CardContent>
                            </Grid>
                            <Grid item xs={4}>
                                <CardContent>
                                    <Typography variant="h4">
                                        {currentTime.hours} : {currentTime.minutes}
                                    </Typography>
                                    <Typography variant="h5">
                                        {currentTime.month} {currentTime.date}
                                    </Typography>
                                    <Typography variant="h6">
                                        {currentTime.year}
                                    </Typography>
                                    <Typography variant="h5">
                                        {currentTime.day}
                                    </Typography>
                                </CardContent>
                            </Grid>
                            <Grid item xs={4}>
                                <CardContent>
                                    <Typography variant="h5">
                                        {locationName}
                                    </Typography>
                                    <Typography>
                                        High {maxTemperature}° | Low {minTemperature}°
                                    </Typography>
                                    <Typography>
                                        Humidity: {humidity}% 
                                    </Typography>
                                    <Typography>
                                        Pressure: {pressure} hPa 
                                    </Typography>
                                </CardContent>
                            </Grid>
                            <Grid item xs={4}>
                                <CardContent>
                                    <Typography variant="h6">
                                        Sun
                                    </Typography>
                                    <Typography>
                                        Sunrise: {sunRiseTime.hours} {sunRiseTime.minutes}
                                    </Typography>
                                    <Typography>
                                        Sunset: {sunSetTime.hours} {sunSetTime.minutes}
                                    </Typography>
                                </CardContent>
                            </Grid>
                            <Grid item xs={4}>
                                <CardContent>
                                    <Typography variant="h6">
                                        Wind Data
                                    </Typography>
                                    <Typography>
                                        Speed: {windSpeed} m/s
                                    </Typography>
                                    <Typography>
                                        Degree: {windDegree}° 
                                    </Typography>
                                </CardContent>
                            </Grid>
                            <Grid item xs={4}>
                                <CardContent>
                                    <Typography variant="h6">
                                        Coordinates
                                    </Typography>
                                    <Typography>
                                        Latitude: {latitute}°
                                    </Typography>
                                    <Typography>
                                        Longitude: {longitude}°
                                    </Typography>
                                </CardContent>
                            </Grid>
                            <Grid item xs>
                            <CardActions>
                                <Link href={`/weather/${latitute}_${longitude}`}>
                                    <Button size="small" color="primary">Get Detailed Weather Forecast</Button>
                                </Link>
                            </CardActions>
                            </Grid>
                        </Grid>
                    </Card>
                </Grid>
            </Grid>
        </>
    );
};

export default WeatherView;