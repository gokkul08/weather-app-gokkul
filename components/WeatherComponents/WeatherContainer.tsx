/**
 * This page is the weather container that performs the API fetching operation for weather from OpenWeatherMaps
 * It allows us to get the weather data for a location and passes date to the WeatherView component
 */

import React, { useState } from 'react';
import WeatherLocationSearch from './WeatherLocationSearch';
import styled from 'styled-components';
import Grid from '@material-ui/core/Grid';
import WeatherView from './WeatherView';
import Button from '@material-ui/core/Button';

const StyledGrid = styled(Grid)`
        padding-top: 20px;
    `;

export const WeatherButton = styled(Button)`
    margin-top: 8px;
`;

export const WeatherContainer = () => {
    const [locationInput, setLocationInput] = useState('');
    const [weatherData, setWeatherData] = useState(null);

    const handleLocation = (newValue: string) => {
      setLocationInput(newValue);  
    };

    const handleWeatherClick = async () => {
        const response = await fetch(`https://api.openweathermap.org/data/2.5/weather?q=${locationInput}&appid=4bf2d273aa7c6b1998f234d78e609def`);
        const data = await response.json();
        if (data) {
            setWeatherData(data);
        }
    };

    return (
        <>
            <StyledGrid container spacing={1}>
                <Grid item xs>
                    <WeatherLocationSearch
                        inputValue={locationInput} 
                        setInputValue={handleLocation}
                    />
                </Grid>
                <Grid item xs>
                    <WeatherButton variant="contained" color="primary" onClick={handleWeatherClick}>
                        Get Weather
                    </WeatherButton>
                </Grid>
            </StyledGrid>
            
            {
                (weatherData !== null && weatherData.cod !== '404') && 
                <WeatherView 
                    data={weatherData}
                />
            }

            {
                (weatherData !== null && weatherData.cod === '404') && 
                 <div> Weather data not found!</div>
            }
        </>
    );
};

export default WeatherContainer;