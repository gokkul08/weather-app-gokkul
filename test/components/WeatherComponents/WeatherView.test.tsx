import React from 'react'
import { render } from '../../test-utils'
import { WeatherView } from '../../../components/WeatherComponents/WeatherView';

jest.mock("next/image", () => ({ src, alt }) => <img src={src} alt={alt} />)

const mockData = {
    'coord': {
        'lat': 17.3616,
        'lon': 78.4747,
    },
    'dt': 1620652723,
    'main': {
        'feels_like': 308.4,
        'humidity': 49,
        'pressure': 1008,
        'temp': 300,
        'temp_max': 300,
        'temp_min': 300,
    },
    'timezone': 19800,
    'sys': {
        'country': 'IN',
        'sunrise': 1620605764,
        'sunset': 1620605764,
    },
    'weather':
        [{
            'main': 'Clouds',
            'icon': '03n',
        }],
    'wind': {
        'deg': 200,
        'speed': 1.54
    }
}

describe('About page', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(<WeatherView data={mockData}/>, {})
    expect(asFragment()).toMatchSnapshot()
  })
})