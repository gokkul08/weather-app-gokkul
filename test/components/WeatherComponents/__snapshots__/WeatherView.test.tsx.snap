// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`About page matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8"
    >
      <div
        class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <b>
               °F
            </b>
            <span
              class="MuiSwitch-root"
            >
              <span
                aria-disabled="false"
                class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiSwitch-switchBase"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <input
                    class="PrivateSwitchBase-input-4 MuiSwitch-input"
                    type="checkbox"
                    value=""
                  />
                  <span
                    class="MuiSwitch-thumb"
                  />
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </span>
              <span
                class="MuiSwitch-track"
              />
            </span>
            <b>
              °C
            </b>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
          >
            <div
              class="MuiCardContent-root"
            >
              <h3
                class="MuiTypography-root MuiTypography-h3"
              >
                80°
              </h3>
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                Feels like 
                <b>
                  95°
                </b>
              </p>
              <img
                src="https://openweathermap.org/img/w/03n.png"
              />
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
          >
            <div
              class="MuiCardContent-root"
            >
              <h4
                class="MuiTypography-root MuiTypography-h4"
              >
                18 : 48
              </h4>
              <h5
                class="MuiTypography-root MuiTypography-h5"
              >
                May 10
              </h5>
              <h6
                class="MuiTypography-root MuiTypography-h6"
              >
                2021
              </h6>
              <h5
                class="MuiTypography-root MuiTypography-h5"
              >
                Monday
              </h5>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
          >
            <div
              class="MuiCardContent-root"
            >
              <h5
                class="MuiTypography-root MuiTypography-h5"
              >
                undefined, IN
              </h5>
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                High 80° | Low 80°
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                Humidity: 49%
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                Pressure: 1008 hPa
              </p>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
          >
            <div
              class="MuiCardContent-root"
            >
              <h6
                class="MuiTypography-root MuiTypography-h6"
              >
                Sun
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                Sunrise: 05 46
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                Sunset: 05 46
              </p>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
          >
            <div
              class="MuiCardContent-root"
            >
              <h6
                class="MuiTypography-root MuiTypography-h6"
              >
                Wind Data
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                Speed: 1.54 m/s
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                Degree: 200°
              </p>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
          >
            <div
              class="MuiCardContent-root"
            >
              <h6
                class="MuiTypography-root MuiTypography-h6"
              >
                Coordinates
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                Latitude: 17.3616°
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                Longitude: 78.4747°
              </p>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
          >
            <div
              class="MuiCardActions-root MuiCardActions-spacing"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  Get Detailed Weather Forecast
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
